File: db/va_releases.py
from db import Database
import sqlite3

class VAReleasesDB(Database):

    def __init__(self):
        super().__init__()
        self.create()

    def create(self):
        self.execute('''CREATE TABLE IF NOT EXISTS vareleases
                    (ID INTEGER PRIMARY KEY AUTOINCREMENT,
                     UserID INTEGER NOT NULL,
                     UserName VARCHAR(100) NOT NULL,
                     ReleaseTitle VARCHAR(100) NOT NULL,
                     ReleaseDate VARCHAR(12) NOT NULL,
                     Desc VARCHAR(300) NOT NULL,
                     Type VARCHAR(10) NOT NULL,
                     Link VARCHAR(200) NOT NULL)''')
        
    def add(self, userID, userName, releaseTitle, releaseDate, desc, type, link):
        self.conn = sqlite3.connect("chh.db")
        query = f"INSERT INTO vareleases (UserID, UserName, ReleaseTitle, ReleaseDate, Desc, Type, Link) VALUES ({userID},\"{userName}\",\"{releaseTitle}\",\"{releaseDate}\",\"{desc}\",\"{type}\", \"{link}\")"
        cursor = self.conn.execute(query)
        self.conn.commit()
        return cursor.lastrowid
    
    def get_user_by_id(self, ID):
        data = self.execute(f"SELECT UserID FROM vareleases WHERE ID={ID}")
        return(data)
    
    def get_by_id(self, ID):
        data = self.execute(f"SELECT * FROM vareleases WHERE ID={ID}")
        return data
    
    def get_by_user(self, userID):
        data = self.execute(f"SELECT * FROM vareleases WHERE UserID={userID}")
        return data
    
    def get_by_date(self, date):
        data  = self.execute(f"SELECT * FROM vareleases WHERE ReleaseDate='{date}'")
        return data
    
    def check(self, ID):
        data = self.execute(f"SELECT UserID FROM vareleases WHERE ID={ID}")
        if len(data) == 0:
            return 0
        else:
            return data[0][0]
    
    def delete(self, ID):
        self.execute(f"DELETE FROM vareleases WHERE ID={ID}")
    
    def update(self, ID, releaseTitle="", releaseDate="", desc="", type="", link=""):
        if not releaseTitle == "":
            self.execute(f"UPDATE vareleases SET ReleaseTitle='{releaseTitle}' WHERE ID={ID}")
        if not releaseDate == "":
            self.execute(f"UPDATE vareleases SET ReleaseDate='{releaseDate}' WHERE ID={ID}")
        if not desc == "":
            self.execute(f"UPDATE vareleases SET Desc='{desc}' WHERE ID={ID}")
        if not type == "":
            self.execute(f"UPDATE vareleases SET Type='{type}' WHERE ID={ID}")
        if not link == "":
            self.execute(f"UPDATE vareleases SET Link='{link}' WHERE ID={ID}")

File: db/vareleases.js
const db = require('./database');
const logger = require('../utility/logger');

class VAReleasesDB {

  async create() {
    logger.info('Checking/creating vareleases table');
    await db.execute(`CREATE TABLE IF NOT EXISTS vareleases
            (ID INTEGER PRIMARY KEY AUTOINCREMENT,
            UserID TEXT NOT NULL,
            UserName VARCHAR(100) NOT NULL,
            ReleaseTitle VARCHAR(100) NOT NULL,
            ReleaseDate VARCHAR(12) NOT NULL,
            Desc VARCHAR(300) NOT NULL,
            Type VARCHAR(10) NOT NULL,
            Link VARCHAR(200) NOT NULL)`);
  }

  async add(userId, userName, releaseTitle, releaseDate, desc, type, link) {

    await this.execute(`INSERT INTO vareleases 
            (UserID, UserName, ReleaseTitle, ReleaseDate, Desc, Type, Link) 
            VALUES 
            ("${userId}","${userName}","${releaseTitle}","${releaseDate}","${desc}","${type}","${link}")`);
    const data = await this.execute(`SELECT ID FROM vareleases 
            WHERE 
            UserID="${userId}" AND UserName="${userName}" AND ReleaseTitle="${releaseTitle}" AND ReleaseDate="${releaseDate}" AND Desc="${desc}" AND Type="${type}" AND Link="${link}"`);
    logger.info(`${userName} added a new release ID #${data[0].ID}, title: ${releaseTitle}, date: ${releaseDate}, desc: ${desc}, type: ${type}, link: ${link} to the vareleases table`);
    return data[0].ID;
  }

  async getUserByID(ID) {
    logger.info(`Getting userID connected to release ID #${ID} in the vareleases table`);
    const data = await this.execute(`SELECT UserID FROM vareleases WHERE ID=${ID}`);
    return data;
  }

  async getByID(ID) {
    logger.info(`Getting the release at ID #${ID} on the vareleases table`);
    const data = await this.execute(`SELECT * FROM vareleases WHERE ID=${ID}`);
    return data;
  }

  async getByUser(userId) {
    logger.info(`Getting all releases for user ID #${userId} from vareleases table`);
    const data = await this.execute(`SELECT * FROM vareleases WHERE UserID=${userId}`);
    return data;
  }

  async getByDate(date) {
    logger.info(`Getting all releases releasing on ${date} from vareleases table`);
    const data = await this.execute(`SELECT * FROM vareleases WHERE ReleaseDate="${date}"`);
    return data;
  }

  async check(ID) {
    logger.info(`Checking that ID #${ID} exists in the vareleases table`);
    const data = await getUserByID(ID);
    return db.checkLen(data);
  }

  async delete(ID) {
    logger.info(`Removing release ID #${ID} from vareleases table`);
    await this.execute(`DELETE FROM vareleases WHERE ID=${ID}`);
  }

  async update(ID, releaseTitle = '', releaseDate = '', desc = '', rType = '', link = '') {
    if (!releaseTitle == '') {
      logger.info(`Updated title on release ID #${ID} to ${releaseTitle} in the vareleases table`);
      await this.execute(`UPDATE vareleases SET ReleaseTitle="${releaseTitle}" WHERE ID=${ID}`);
    }
    if (!releaseDate == '') {
      logger.info(`Updated release date on release ID #${ID} to ${releaseDate} in the vareleases table`);
      await this.execute(`UPDATE vareleases SET ReleaseDate="${releaseDate}" WHERE ID=${ID}`);
    }
    if (!desc == '') {
      logger.info(`Updated desc on release ID #${ID} to ${desc} in vareleases table`);
      await this.execute(`UPDATE vareleases SET Desc="${desc}" WHERE ID=${ID}`);
    }
    if (!rType == '') {
      logger.info(`Updated type on release ID #${ID} to ${rType} in vareleases table`);
      await this.execute(`UPDATE vareleases SET TYPE="${rType}" WHERE ID=${ID}`);
    }
    if (!link == '') {
      logger.info(`Updated link on release ID #${ID} to ${link} in vareleases table`);
      await this.execute(`UPDATE vareleases SET LINK="${link}" WHERE ID=${ID}`);
    }
  }
}

module.exports = VAReleasesDB;